File: netgo/src/main/java/com/fungo/netgo/utils/NetLogger.java
package com.fungo.netgo.utils;

import android.util.Log;

/**
 * @author Pinger
 * @since 18-08-17 上午9:43
 * NetGo框架的日志打印
 */
public class NetLogger {

    private static boolean isLogEnable = true;

    private static String tag = "NetGo";

    public static void debug(boolean isEnable) {
        debug(tag, isEnable);
    }

    public static void debug(String logTag, boolean isEnable) {
        tag = logTag;
        isLogEnable = isEnable;
    }

    public static boolean isDebug() {
        return isLogEnable;
    }

    public static void v(String msg) {
        v(tag, msg);
    }

    public static void v(String tag, String msg) {
        if (isLogEnable) Log.v(tag, msg);
    }

    public static void d(String msg) {
        d(tag, msg);
    }

    public static void d(String tag, String msg) {
        if (isLogEnable) Log.d(tag, msg);
    }

    public static void i(String msg) {
        i(tag, msg);
    }

    public static void i(String tag, String msg) {
        if (isLogEnable) Log.i(tag, msg);
    }

    public static void w(String msg) {
        w(tag, msg);
    }

    public static void w(String tag, String msg) {
        if (isLogEnable) Log.w(tag, msg);
    }

    public static void e(String msg) {
        e(tag, msg);
    }

    public static void e(String tag, String msg) {
        if (isLogEnable) Log.e(tag, msg);
    }

    public static void printStackTrace(Throwable t) {
        if (isLogEnable && t != null) t.printStackTrace();
    }

}

File: netgo/src/main/java/com/fungo/netgo/utils/NetLogger.kt
package com.fungo.netgo.utils

import android.util.Log

import org.json.JSONArray
import org.json.JSONObject

/**
 * Author  ZYH
 * Date    2018/1/24
 * Des     NetLogger
 */
object NetLogger {
    var DEBUG = true/*BuildConfig.DEBUG*///这个待定写出根据debug模式来
    val TAG = "fungo_net_test"
    private val MIN_STACK_OFFSET = 3
    private val METHOD_COUNT = 1
    private val JSON_INDENT = 4
    /**
     * Android's max limit for a log entry is ~4076 bytes,
     * so 4000 bytes is used as chunk size since default charset
     * is UTF-8
     */
    private val CHUNK_SIZE = 4000

    @Synchronized private fun log(logType: Int, str: String, showHeader: Boolean) {
        var content = str
        if (DEBUG) {
            if (showHeader) {
                logHeaderContent()
            }

            content = "| " + content
            val bytes = content.toByteArray()
            val length = bytes.size

            var i = 0
            while (i < length) {
                val count = Math.min(length - i, CHUNK_SIZE)
                val msg = String(bytes, i, count)

                when (logType) {
                    Log.ERROR -> Log.e(TAG, msg)
                    Log.INFO -> Log.i(TAG, msg)
                    else -> Log.v(TAG, msg)
                }
                i += CHUNK_SIZE
            }
        }
    }

    private fun log(logType: Int, content: String) {
        log(logType, content, true)
    }

    fun i(content: String) {
        log(Log.INFO, content)
    }

    @JvmStatic
    fun i(args: Array<String>) {
        try {
            val stringBuilder = StringBuilder()
            for (arg in args) {
                stringBuilder.append(arg)
                stringBuilder.append(",")
            }
            i(stringBuilder.toString())
        } catch (throwable: Throwable) {
            e(throwable)
        }

    }

    fun i(content: Int) {
        i(content.toString())
    }

    fun d(tag: String, content: String) {
        if (DEBUG) {
            Log.d(tag, content)
        }
    }

    fun e(content: String) {
        log(Log.ERROR, content)
    }

    fun e(throwable: Throwable) {
        var msg = ""
        if (throwable.message != null) {
            msg = throwable.message!!
        }

        e(msg)
        throwable.printStackTrace()
    }

    fun json(json: String?) {
        if (DEBUG) {
            try {
                var message = json
                if (json == null || json.isEmpty()) {
                    message = ""
                } else if (json.startsWith("{")) {
                    val jsonObject = JSONObject(json)
                    message = jsonObject.toString(JSON_INDENT)
                } else if (json.startsWith("[")) {
                    val jsonArray = JSONArray(json)
                    message = jsonArray.toString(JSON_INDENT)
                }
                i(message!!)
            } catch (throwable: Throwable) {
                var msg = ""

                if (throwable.cause != null && throwable.cause!!.message != null) {
                    msg = throwable.cause!!.message!!
                }

                e(msg + "\n" + json)
            }

        }
    }

    private fun logHeaderContent() {
        try {
            val trace = Thread.currentThread().stackTrace
            val stackOffset = getStackOffset(trace)

            var stackEnd = stackOffset + METHOD_COUNT
            if (stackEnd >= trace.size) {
                stackEnd = trace.size - 1
            }

            //Log.i(TAG, "|------------------------------------------------------");
            Log.d(TAG, "|---Thread: " + Thread.currentThread().name + " ------------------------------------------------------")

            for (i in stackEnd downTo stackOffset) {

                val builder = StringBuilder()
                builder.append("| ")
                        .append(getSimpleClassName(trace[i].className))
                        .append(".")
                        .append(trace[i].methodName)
                        .append(" ")
                        .append(" (")
                        .append(trace[i].fileName)
                        .append(":")
                        .append(trace[i].lineNumber)
                        .append(")")
                Log.d(TAG, builder.toString())
            }
        } catch (throwable: Throwable) {
            e(throwable)
        }

    }

    private fun getStackOffset(trace: Array<StackTraceElement>): Int {
        var start = -1

        for (i in MIN_STACK_OFFSET until trace.size) {
            val element = trace[i]
            val name = element.className

            if (name != NetLogger::class.java.name) {
                start = i
                break
            }
        }
        return start
    }

    private fun getSimpleClassName(name: String): String {
        val lastIndex = name.lastIndexOf(".")
        return name.substring(lastIndex + 1)
    }
}
